### ISSUE TITLE
Comments: Replacing Disqus with Github Comments

### ISSUE THREAD

**Issue by dwilliamson** [dwilliamson/donw.io#1]
Comments for [Replacing Disqus with Github Comments](http://donw.io/post/github-comments/)

**Comment by dwilliamson** [comment]
This is a comment

**Comment by dwilliamson** [comment]
This is another comment

**Comment by dwilliamson** [comment]
This is a comment with some markdown...

* Listy
* Listo

1. Jumble
1. Sale

**Boldit** or *Italicit* or ~strikeit~

header 
---

## hashtag

something `something`

```cpp
int x = 3;
```

**Comment by jasminpatry** [comment]
This is great. Thanks for sharing this!

**Comment by olafleur** [comment]
Awesome ! I think I will try this on my website. :)

**Comment by vielmetti** [comment]
Nicely done.

**Comment by vyp** [comment]
Neat!

**Comment by olalonde** [comment]
You are on HN, congrats :) https://news.ycombinator.com/item?id=14170041

**Comment by mannuscript** [comment]
I agree that what Disqus is doing is an overkill.
However, we should not forget that nothing comes for free...
Apart from this, getting redirected to GitHub for comment is very inconvenient...
Update: So I am automatically subscribed to comments thread via email...
Getting emails not only for replies on my comment but for every new comment.
Well, this is another overkill...
Don't forget User experience is more valued than user's data nowadays :D 

Edit: 
@dwilliamson forget about the free-ness argument, the bigger issue is of user experience, I got resubscribed to email updates after you mentioned me in one of the comment. This is a serious poor user experience.

**Comment by rattrayalex** [comment]
Interesting, thanks for sharing! 

You might consider changing the link to include `#new_comment_field` ([like so](https://github.com/dwilliamson/donw.io/issues/1/#new_comment_field)) so that users do not have to scroll. I believe this also focuses the field by default, which may make intent more clear. 

Adding a button at the bottom of the comments section that opens the link might be helpful. Using `about: _blank` could be nice too. 

A git hook might be useful to automatically generate the issues, though I'm not sure that'd be a good idea. 

EDIT: Looks like their api supports POSTing comments; you'd have to set up user auth of course. https://developer.github.com/v3/issues/comments/#create-a-comment

**Comment by ghost** [comment]
A test of comment with image and link.

[![img_4698](https://cloud.githubusercontent.com/assets/951921/25300522/17fee0d8-26df-11e7-9588-8674b7d3b81a.PNG)](https://github.com/dwilliamson/donw.io/commit/d7f1f5874d704b2d68a90139e14f4a8d247e2f7b#commitcomment-21864198)

**Comment by globalcitizen** [comment]
Good work. In summary, to improve the solution further, add a button to add a new comment. Only when that button is pushed (in order to preserve privacy, remove drive-by viewer tracking potential, and reduce unnecessary server load on Github), somehow determine whether the user is already logged in to Github.
  - If not, then give them a chance to do so or sign up
  - If so, then display an add comment form on the page
This could be a modal window, in a similar style to Stripe. If the user posts a comment successfully, the comment stream could be refreshed.

**Comment by ghost** [comment]
I'm impressed by this! I'm going to see what happens when you post a comment from the @ghost account (another name for a deleted account)

**Comment by ghost** [comment]
Hello world! This is a comment from the @ghost account.

**Comment by segiddins** [comment]
This is awesome!

**Comment by motyar** [comment]
Ok You know I am testing. testing a link this time! https://gulpx.com

**Comment by olalonde** [comment]
You might want to

1) Add `rel="nofollow"` to your anchor tags to deter spam

2) Link back to the blog post in the GH issue (you could do that in the first comment and not display it on your website). That way when I go back here due to :octocat: GH notifications, I can go back to the post easily instead of wondering what this is all about 😸

**Comment by ElijahLynn** [comment]
Great writeup, I was really hoping there was a way to comment without visiting Github.

**Comment by asciimo** [comment]
Very slick! I came to propose something similar to @globalcitizen's proposal, but they put it more succinctly.

**Comment by bexelbie** [comment]
Did you consider https://staticman.net ? This gets you comments as PRs and comments from the page.

**Comment by tbodt** [comment]
@bexelbie but staticman isn't as FUCKING AWESOME as this

**Comment by dipakc** [comment]
This is amazing! Thanks for sharing.

**Comment by aleman** [comment]
Great! 
I can already think of a few use cases for this in my team

**Comment by colinmegill** [comment]
Great idea. Thank you for this.

**Comment by crohr** [comment]
I agree this is a great idea. But why not write your post directly into the issue as well? You might want to checkout my post [Turn your GitHub issues into blog posts](http://ebarnouflant.com/posts/4-turn-your-github-issues-into-blog-posts), which implements the same idea but also let you publish right from your GitHub issue as well.

**Comment by bkaradzic** [comment]
Love it! :+1:

**Comment by lrzedzicki** [comment]
Very useful!

**Comment by andreis** [comment]
Great technique that I'll definitely steal. Thanks!

**Comment by ghost** [comment]
Brilliant idea!

**Comment by EddieOne** [comment]
The best idea kind of. Follow me 😀  https://onefoldmedia.com

**Comment by EddieOne** [comment]
``` javascript
test = {
  something: 'monkey',
  function: function(arg) { return arg + 'YAY!'  }
}
```

Edit: We need pagination!

**Comment by cerisara** [comment]
Great ! I'll try this out... is it doable with a gitlab instance ?

**Comment by spinda** [comment]
![](http://i.imgur.com/X2iQxpQ.png)

**Comment by export-mike** [comment]
Super clever!

**Comment by thehesiod** [comment]
note I filed a bug LAST MONTH where I found that the github API was returning the same comments multiple times instead of all the comments for large number of comments (500+).  They have yet to say they've fixed this.  So for example if you had comments: [1,2,3,4], it would return [1,1,3,4]

**Comment by spinda** [comment]
![test](https://trap.spinda.net/trap.png)

**Comment by modest** [comment]
cute, but... when we start needing to solve a Captcha before commenting on GitHub, I'll be sure to thank you guys

**Comment by rezhajulio** [comment]
did it also prevent executing js ? <script>alert("HA!")</script>

**Comment by DanTup** [comment]
Worth noting, the [API results](https://api.github.com/repos/dwilliamson/donw.io/issues/1/comments) for this comment currently appear to be 1hr stale (probably due to GH caching) which could be important to some in deciding whether to do this.

**Comment by DanTup** [comment]
Actually, they're not stale, they're paged?

https://api.github.com/repos/dwilliamson/donw.io/issues/1/comments
https://api.github.com/repos/dwilliamson/donw.io/issues/1/comments?page=2

**Comment by mtso** [comment]
I wonder if it's possible to post from your blog page~
Cool implementation tho~

**Comment by nicolas-besnard** [comment]
kjkj

**Comment by tylerzika** [comment]
really cool! although I'm not seeing this on your blog post..

**Comment by kennycason** [comment]
Woah! Awesome.

```kotlin
println("Everyone switch to Kotlin")
```

**Comment by rhaps0dy** [comment]
I'm going to implement this in my blog. Thank you 😄 .

**Comment by DmitryOlkhovoi** [comment]
It's better to post comments via api than ask ppl to do it here

**Comment by rhaps0dy** [comment]
@DmitryAuine True, the notice "go to the GitHub issue to comment" is easy to miss. Well we can try to implement this. (I might at some point)

**Comment by 0b01** [comment]
Check out txtpen.com. It is like medium but for regular sites

**Comment by elliotblackburn** [comment]
Huh awesome, might add this into my Ghost blog!

**Comment by intraz** [comment]
@dwilliamson, what browser did you use for the network log screen captures?
thanks for a good post.

**Comment by milch** [comment]
@DanTup It looks like if the comments are paged then the navigation is included in the `Link` response header like this:

```
<https://api.github.com/repositories/72930823/issues/1/comments?page=1>; rel="first", <https://api.github.com/repositories/72930823/issues/1/comments?page=1>; rel="prev"
```

So it'd be neat if the JS in the `comments.html` file detected if the user scrolled to the bottom, and then loaded the next page of comments if there are any.

**Comment by pwnlambda** [comment]
<script>alert(1)</script> Testing for input sanitizing

**Comment by tennisfar** [comment]
Simple and efficient. Good idea. 👍

**Comment by jfernandez** [comment]
Hi mom!

**Comment by AbhishekBiswal** [comment]
Testing. [Matchr](https://matchr.in)

**Comment by Aprillion** [comment]
@dwilliamson github api has a limit of 30 items by default - https://developer.github.com/v3/#pagination

compare:
* https://api.github.com/repos/dwilliamson/donw.io/issues/1/comments
* https://api.github.com/repos/dwilliamson/donw.io/issues/1/comments?per_page=100

**Comment by Turbots** [comment]
Very nicely done, we will check out how we can leverage this on our tech blog!

**Comment by danshearmur** [comment]
<img src="lol.jpg" onerror=alert('woah')>

**Comment by mahadevTW** [comment]
Cant i Use github issuie Api to comment from your site , so that to just comment on you post i need not to navigate to github.com?

**Comment by aemxn** [comment]
Testing testing one two three

**Comment by lukaskollmer** [comment]
nice

**Comment by lukaskollmer** [comment]
this is an awesome idea btw

**Comment by endel** [comment]
Well done. 👏

**Comment by blockjon** [comment]
foo

**Comment by jlucasps** [comment]
Nice :-)

**Comment by boramalper** [comment]
How to import comments from Disqus though, any suggestions?

**Comment by fasiha** [comment]
It'd be nicer (for some blogs) if the comments were stored in the git repo for easy export out of GitHub, for when GitHub eventually dies 😿.

**Comment by ghost** [comment]
I really like this. GitHub should see how far they can take it!

**Comment by guoliim** [comment]
nice idea

**Comment by JamesKim2998** [comment]
new comment

**Comment by styfle** [comment]
Does the github API support adding a comment to an issue? If yes, then it would make sense to allow the user to add a comment without leaving the original blog. 

As others have pointed out, the blog post should be the content of the Issue so that it makes sense what users are commenting on.

**Comment by akshaymohite** [comment]
This poses an issue if we want to have nested comments. Any thoughts anyone? @dwilliamson

**Comment by dmitshur** [comment]
> Does the github API support adding a comment to an issue? If yes, then it would make sense to allow the user to add a comment without leaving the original blog.

It does, but it requires authentication.

**Comment by ouvek-kostiva** [comment]
Awesome!
And never knew that disqus used so much tracking.

**Comment by chrisbrandow** [comment]
That's clever. no threading, but `¯\_(ツ)_/¯`

**Comment by salehhamadeh** [comment]
This is the most creative idea I have seen in a **while**.

**Comment by stefan-it** [comment]
🥇

**Comment by augbog** [comment]
Look mom I made it.

In all seriousness though great post :)

**Comment by chris124567** [comment]
Cool stuff!

**Comment by vy** [comment]
Neat solution! Thanks for sharing.

**Comment by globalcitizen** [comment]
@akshaymohite Nested comments could be done like this, where a second pass orients children with leading '@ + username' beneath the immediately prior post by the referenced user.

```
user A: "Today a young man on acid realized that all matter is merely energy condensed to a slow vibration, that we are all one consciousness experiencing itself subjectively, there is no such thing as death, life is only a dream, and we are the imagination of ourselves. Heres Tom with the Weather."
user B: "@A Sunny with a chance of a late afternoon satellite-disabling solar storm."
user A: "@B You don't need a GPS to know which way the solar wind blow."
user B: "@A We're going deep now..."
```

The above would become a single tree as follows:
```
A: "Today a young man on acid..."
 |
 `- B: "Sunny with a chance..."
      |
      `- A: "You don't..."
           |
           `- B: "We're going deep now..."
```

From an algorithmic efficiency standpoint assigning a numeric identifier to each post and doing a single pass (while maintaining a 'last seen post by user' structure with which to allocate 'parent' for each post) would suffice.

## Responding to earlier comments

### Quotation method
If additional functionality was required (ie. the ability to respond to very early comments) then quotations could be used:

```
user A: @B > Sunny with a chance...
What are we talking here, statistically?
```

(ie. `@B` identifies the user, `>` starts the quote, and the rest of the comment is the actual content. This would require pattern-matching to work reliably which is more computationally expensive than the above algorithm. Using it only when required would presumably increase readability.)

### Numeric method

Alternatively, a numbered post reference could be used, since Github apparently numbers them already (eg. URLs like https://github.com/dwilliamson/donw.io/issues/1#issuecomment-296385911). In this case, quoting precisely the post ID could be useful, though it would more visual pollution and less semantic context for the casual reader it would prevent the need to rebuild the comments tree in a computationally expensive fashion on every view. (Alternatively, a server precomputed/cached model of the quotation method could also be used to mitigate the overhead.)

**Comment by odan** [comment]
Gist (from Github) has a nice comment system too :-)

The comments for a gist post can be read as JSON. Example:
https://api.github.com/gists/1abe76d373a9cbb15bed/comments

The Gists API: https://developer.github.com/v3/gists/

**Comment by alecstanescu** [comment]
Apparently it is hard to comment on this issue from a mobile browser, so I made a mistake and opened another issue. Anyway great work

**Comment by eden-lane** [comment]
On Russian site frontender.info you can even login with github and leave comments from the site:
https://frontender.info/document-outlines-in-html-5-1/

**Comment by dmitshur** [comment]
> Gist (from Github) has a nice comment system too :-)

@odan, the main downside, compared to issue comments, is that gist comments don't have any kind of notification system.

**Comment by pjar** [comment]
Again - this is super clever! :) 
I didn't read all the comments (guys are too enthusiastic! :D ) and probably someone already wrote that, but  further steps could be:
- 1 comment = 1 issue - that way it could virtually make it possible to have threads
- utilize GitHub API to embed comment forms on the site

**Comment by DanTup** [comment]
> utilize GitHub API to embed comment forms on the site

I suspect this wouldn't be possible for a "static" site without having some secrets in the visible source code. It'f probably need supporting server-side code (and if you have the ability to do this, you'd probably just put your own comments form in and avoid GH entirely?).

**Comment by fadzlan** [comment]
I think Disqus is visiting those sites because those are the identity owners of their users. That means if you are already logged in to Facebook, you can comment as the identity of your Facebook account.

I don't see that's going away if you are going to support multiple identity providers. On the other hand, if you are just engaging with only the developer communities, I can see whey Disqus is overkill.

**Comment by zhaochunqi** [comment]
check this one [https://github.com/imsun/gitment.git](https://github.com/imsun/gitment.git)

**Comment by Jasonnor** [comment]
Cool work!

**Comment by adoyle-h** [comment]
So great! 😺

**Comment by cyberwani** [comment]
Awesome 👍

**Comment by ahmadalfy** [comment]
LMAO

**Comment by Tifox** [comment]
trytry

**Comment by jeffersonbsilva** [comment]
Great! @dwilliamson

**Comment by odan** [comment]
@shurcooL https://giscus.co/ provides a comment notification system gists. (Not tested yet)

**Comment by dwilliamson** [comment]
Thank you, everybody, for all your comments and suggestions of alternatives. My blog is pretty niche and rarely generates any comment threads so there's some irony in this particular post exploding. Some answers below...

@rattrayalex @globalcitizen thanks, the button is an easy addition that would improve posting (minus oauth).

@mannu1200 well aware of the free-ness argument but it does not justify the by-default privacy invasion.

@rattrayalex @DmitryAuine @mahadevTW @styfle Right now I don't want to require auth or implement the required forms. Github does it well enough for my needs. I'm not trying to build a community, just give people a chance to query my posts. Not to mention that as a long-time Github user myself, I very rarely give any service login authorization.

@ontouchstart good point, I should put the code in a script file

@olalonde yup, totally intended to do number 2 but forgot.

@bexelbie never seen that one but with a few modifications, this is all I need for now. Thanks for the link.

@crohr I mentioned that in my post but I don't want to be bound to Github for actual site content.

@EddieOne @DanTup @tylerzika only after you guys posted comments did I realise it was page-limited! Will add that soon.

@DanTup @Aprillion thanks for figuring out the page parameters.

@cerisara no idea.

@modest this site is open-source and all issues are opened on content hosted in a Github repo. Access to the API JSON is open as a conscious design choice and rate limited. If a spam-bot can follow links from a page that simply present issue comments as issue comments then it's just as capable of following links that present issue comments as post comments. All of this is within the TOC and encouraged by the API design.

@rezhajulio Github sanitizes the markdown so that we don't have to worry about that.

@intraz rendering was done with http://www.webpagetest.org/

@boramalper sorry, no.

@fasiha cute idea. You can crawl the JSON and commit for now, I guess.

@akshaymohite I'm not too bothered as I think being able to reference somebody by handle and quote is good enough for me.

@odan @zhaochunqi thanks for the alternatives.

@fadzlan no, this is just as bad without being logged into facebook or google. You are being tracked and your movements are being sold to unknown 3rd parties as part of the Disqus business model.

**Comment by intraz** [comment]
@dwilliamson, thanks for the answer

**Comment by boxcore** [comment]
cool staff!

**Comment by christophberger** [comment]
Thanks for creating and sharing this. That's really a creative way of using GitHub Isssues, hats off!

 I have enabled Disqus on my blog after resisting for many months (for exactly the same reasons stated in this article), but then I decided to add Disqus as I found no suitable alternative. Finally, it seems I have found an alternative; I hope I can somehow migrate the few comments that were already posted.

EDIT: And +1 for being Hugo-specific!

**Comment by revant117** [comment]
this is a comment

**Comment by Esqarrouth** [comment]
Cool piece

**Comment by lukeramsden** [comment]
This is nuts

**Comment by EddieOne** [comment]
Ah someone deleted my link. Got a little taste of that Hacker News traffic anyway. 😈

**Comment by csi-lk** [comment]
Agreed, pagination is a good idea

**Comment by Hydrock** [comment]
Test

**Comment by vkopitsa** [comment]
![image](https://cloud.githubusercontent.com/assets/1539559/25431132/f623e330-2a87-11e7-84db-d105226b8955.png)

**Comment by ivanbalashov239** [comment]
this solution doesn't have tree view comments.
I think disqus's mast have feature is tree view comments.

**Comment by ogroleg** [comment]
> quote 

- test

**Comment by cetver** [comment]
aaa

**Comment by Dragomeat** [comment]
Test

**Comment by dra1n** [comment]
Nice hack

**Comment by undiabler** [comment]
test js inject
<script>alert('hacked!');</script>

**Comment by englishextra** [comment]
Much noise, no working alternative solution. Or, there is?

**Comment by sklyar** [comment]
temp

**Comment by isatimur** [comment]
Why there is no iframe window to have a direct access to text you some comments directly from your post?!?

**Comment by samarkandiy** [comment]
This is nice post!

**Comment by Pycz** [comment]
Wow!

**Comment by ghost** [comment]
To create discussion thread, you do not have to use permanent GitHub issues. [gists](https://help.github.com/articles/about-gists/) also allow comments . Here is an experiment thread for you to play with: [A light weight non-public gist based discussion thread](https://gist.github.com/ontouchstart/1fc533578feb582980a5be25e76a9920).

**Comment by eks1985** [comment]
Nice solution

**Comment by shegeley** [comment]
Noice! 
![hqdefault](https://cloud.githubusercontent.com/assets/8594472/25475230/be75ea98-2b3e-11e7-8223-f054b7e9db1c.jpg)

**Comment by makc** [comment]
тесто

**Comment by makc** [comment]
some options to scroll this page down to the input form:

https://github.com/dwilliamson/donw.io/issues/1#new_comment_field

https://github.com/dwilliamson/donw.io/issues/1#partial-new-comment-form-actions

also you can use github api to implement comment box directly on your site:

https://developer.github.com/v3/issues/comments/#create-a-comment

**Comment by zelds** [comment]
So try to test it.

**Comment by jpuck** [comment]
@cerisara unfortunately at this time [Gitlab issues API][1] would require you to expose your key.

> Every API call to issues must be authenticated.

But you could always host the issues on Github while the site is on Gitlab.

[1]:https://docs.gitlab.com/ee/api/issues.html

**Comment by jpuck** [comment]
@dwilliamson It looks like the limit is 30 comments to display. Any idea how to paginate, or at least get all the comments? The [API docs][1] show the `since` param, but that would only allow you to show the most recent.

[1]:https://developer.github.com/v3/issues/comments/

**Comment by dwilliamson** [comment]
@jpuck thread mentions how to request issue comments beyond the initial page https://developer.github.com/guides/traversing-with-pagination/

**Comment by chibani** [comment]
This is simply great !

**Comment by dwilliamson** [comment]
I've added pagination and new comment field scroll to the latest if anybody's interested. There are two methods; the first is the typical page select:

![image](https://cloud.githubusercontent.com/assets/1532903/25520988/efe47182-2bf5-11e7-936a-a7a240c3f88c.png)

I found that unwieldy for reasons mentioned in the commit which then simply added a button:

![image](https://cloud.githubusercontent.com/assets/1532903/25521009/fea51816-2bf5-11e7-8417-09adb2c5f5c4.png)

Code for both is ce31f79aab2e62998ec20d44b08d163b190c0683 and 85a0ee449c9f66b312cafebe2864bb40af14f0a3.

It's worth mentioning that the way it's implemented is a hang-over from the original page select so it's manually constructing queries against Github's guidelines. Will fix that up proper when I get a chance - should make the code even simpler.

**Comment by srbraun** [comment]
**It works.** 

> Awesome.

 _## Thank you._

**Comment by greggman** [comment]
Are you sure all those requests you listed are coming from disqus?

Here's page that uses disqus: https://webglfundamentals.org/webgl/lessons/webgl-and-alpha.html

And here's the network tab from chrome

![screen shot 2017-05-02 at 1 38 18 pm](https://cloud.githubusercontent.com/assets/234804/25604345/b0cebfa2-2f3d-11e7-8f16-50443eef8646.png)

I see none of the 3rd party domains shown in your post.

Even going to [a page with many more comments](https://webglfundamentals.org/webgl/lessons/webgl-fundamentals.html) I see no 3rd party domain requests what-so-ever.  

I have ads turned off, maybe that's why?

![screen shot 2017-05-02 at 1 44 31 pm](https://cloud.githubusercontent.com/assets/234804/25604323/85f3444c-2f3d-11e7-9ce2-bac2ef500833.png)

I mean sure if people post picture links there will be 3rd party domain requests but that's no different with github. Here's a direct link to giphy

> ![puppies](https://media.giphy.com/media/gxxlowyvtfS0M/giphy.gif)

This also seems against github policy, if not actual words at least in spirit. The issue system is here is for supporting either open source software projects or private (paid for to github for hosting) software projects. It is not here for comments on your personal blog (unless the blog is open source or paid for on github). It just seems like abuse of their system to use it for random comments. Of course whether they think it is abuse you'd have to ask them. Even worse, you're asking people who have no interest in github to make github accounts. That's both not useful to github themselves (they have no interest in my mom making a github account for comments) and it's not very friendly to users (using disqus they can use existing accounts). Just my 2¢

**Comment by growdigital** [comment]
Sweet idea, thanks. Not sure it would take off for "non-technical" folk but that's not really a deal-breaker, I find disqus awkward enough to use as it is.

**Comment by iLeonidze** [comment]
Is this is safe for the site? How about malware frames/scripts?

**Comment by y6nH** [comment]
Really nice idea, for sites whose visitors are likely to be signed up on Github. I'm going to put this in my Hugo templates.

**Comment by dwilliamson** [comment]
@greggman 
1. Disable blockers.
2. I've answered all of your use of GH queries in earlier posts or in the blog. Please take the time to catch up.

**Comment by greggman** [comment]
All my blockers were disabled. You'll notice in the top picture the icons for ublock origin and adblock don't appear (where as they do in the 2nd).  So, I'm still wondering why I'm getting different results than you got. My guess was maybe your disqus settings are different.

As for policy the only answer I noticed (which I did read before) was the one you made to @ modest and that's fine. I just seems like it would be nice if you put something near the top saying "don't use this unless your website is open source or you're a paid github user".

**Comment by zeetabit** [comment]
hm

**Comment by dwilliamson** [comment]
@greggman that seems like a sensible prefix to add to the post.

**Comment by Zhekaa** [comment]
1111

**Comment by madbob** [comment]
Ages ago I implemented a "UserVoice" alternative using the same concept - http://vh.madbob.org/

**Comment by ddsol** [comment]
@dwilliamson: GitHub can track you across sites with this system using the the `Origin` header.

GitHub can utilize cookies (if 3rd party is enabled) and assign browsing habits to a user, or possibly track by IP.

GitHub might also use this to build aggregate data, which may or may not be on people's privacy radar.

That said, it's not GitHub's business model to track, aggregate and sell data.

**Comment by akodkod** [comment]
lol, this is my comment on site

**Comment by sid3effect** [comment]
For real?

**Comment by geekodour** [comment]
this is cool

**Comment by sid3effect** [comment]
@geekodour no it's not coz you have to leave the original website to post comment

**Comment by Denysria** [comment]
Cool

**Comment by spanishharlem** [comment]
Interesting idea.

**Comment by dwilliamson** [comment]
@ddsol only a JSON data request is made from the site to Github. Your reading habits can not be tracked by Github using this model. As mentioned in the post, IPv4 addresses are not sufficient for reliable tracking due to technology such as CGNAT.

**Comment by Astorsoft** [comment]
Let's try this!

**Comment by dwighthouse** [comment]
Something tells me that Github might not approve of this sort of thing.

**Comment by Vacilando** [comment]
Another test comment

**Comment by dmitshur** [comment]
@dwilliamson, I'm curious if you've seen https://dmitri.shuralyov.com/blog/23 and what you think about it. It's inspired by your post, and demonstrates something similar I've done on my blog.

**Comment by d3netxer** [comment]
nice post! comments should be in reverse chronological order and would be nice to comment directly in blog

**Comment by dwilliamson** [comment]
@shurcooL yes, seen that. It's a little too far for me in that I want to control my blog content; comments aren't a big feature of my site.

@d3netxer Both good points though I may never get around to doing that as the return isn't great for my site.

**Comment by craigbuckler** [comment]
Nice idea. Shame you have to enter the comments at GitHub itself, though.

**Comment by jia3ep** [comment]
Interesting idea. But how to detect spam messages?

**Comment by nyorc** [comment]
i **love** this comment system.

**Comment by berot3** [comment]
really nice way of commenting! :dancing_men:

**Comment by chuckreynolds** [comment]
that's a cool idea. i like 👍

**Comment by husnimun** [comment]
Thanks for sharing this!👍

**Comment by nikialeksey** [comment]
Test comment

**Comment by dc25** [comment]
Another test comment.

**Comment by ghost** [comment]
This is absolutely metal, and I'm jealous that I never thought of doing this myself. Thanks for posting this! :metal:

BTW, I found this post via [r/jekyll](https://www.reddit.com/r/Jekyll/comments/6huuip/using_github_issuescomments_in_jekyll_blog/) on Reddit.

**Comment by dc25** [comment]
Do you mind if I use some of the javascript code that you wrote for this?  Your blog post implies that this would not be a problem but I don't see a license and want to make sure.

**Comment by dwilliamson** [comment]
@dc25 that's fine. Assume MIT license for code and CC BY-NC for posts/articles.

**Comment by dc25** [comment]
I adapted your code to work with Jekyll.   [Here's a write up](https://dc25.github.io/myBlog/2017/06/24/using-github-comments-in-a-jekyll-blog.html).

**Comment by mateuspntx** [comment]
## Really nice!

**Comment by Tsui89** [comment]
good idea.

**Comment by mika-cn** [comment]
This is really awesome and useful,  Disqus can't work properly if reader access from china (due to GFW).
Thank you :+1:

**Comment by ghost** [comment]
This is really cool!

**Comment by davchezt** [comment]
cool

**Comment by isdux** [comment]
a

**Comment by isdux** [comment]
hello, i have a question, i use this in my blog. but i can' t find the parameter about comments_id. i use swing to add the div and script. thanks in advance.

**Comment by isdux** [comment]
i got it,thans.

**Comment by isdux** [comment]
i got it,thans. And right now use in my blog.
[home](http://isdux.com)
This is really cool!

**Comment by metacritical** [comment]
WoW, A novel idea but i would really prefer an embedded comment box in my site.

**Comment by tcgriffith** [comment]
genius

**Comment by nancym** [comment]
I'm not able to get this to work and I'm wondering if anything has changed to make the code in your post not work. I've tried a few things, including adding `<script>` and `</script>` to the top and bottom of `comments.html`. I also tried using the code you have in your repo today, but that seems to use node. I'd like to use vanilla javascript because my hosting provider is DreamHost and I'd like to keep it as vanilla as possible, e.g., not enable node there. Thanks for any tips!

**Comment by dwilliamson** [comment]
All this still works but it's most definitely not using node.js! This entire site is statically generated using Hugo and you need that installed in order to use any of this.

**Comment by dwilliamson** [comment]
All this still works but it's most definitely not using node.js! This entire site is statically generated using Hugo and you need that installed in order to use any of this.

**Comment by VEBERArnaud** [comment]
:camel::dash:

**Comment by 1mike12** [comment]
``` html
<body style="">
<!--[if lt IE 7]><p class="browsehappy">You are using an <strong>outdated</strong> browser. Please <a href="http://browsehappy.com/">upgrade your browser</a> or <a href="http://www.google.com/chrome/‎">install Google Chrome</a> to experience this site.</p><![endif]-->

    <header id="site-header">
        <div class="container">
            <a href="http://donw.io/" alt="Gazoo.vrv"><h1 class="blog-title heading" style="font-size: 30px;">Gazoo.vrv</h1></a>
            <p class="blog-description">Don Williamson, Consultant Game Technology Programmer/Director of Celtoys</p>
			
        	
        	<li class="navigation__item">
            	<a href="http://twitter.com/Donzanoid" title="@Donzanoid on Twitter"> <i class="fa fa-twitter"></i> <span class="label">@Donzanoid</span> </a>
        	</li>
        	
			
        	
        	<li class="navigation__item">
            	<a href="https://github.com/dwilliamson" title="dwilliamson on github"> <i class="fa fa-github"></i> <span class="label">dwilliamson</span> </a>
        	</li>
        	        
            
            
            <li class="navigation__item">
                <a href="https://github.com/Celtoys" title="Celtoys on github"> <i class="fa fa-github"></i> <span class="label">Celtoys</span> </a>
            </li>
                    
			
        	
        	<li class="navigation__item">
            	<a href="https://bitbucket.org/dwilliamson" title="dwilliamson on bitbucket"> <i class="fa fa-bitbucket"></i> <span class="label">dwilliamson</span> </a>
        	</li>
        	
			
        </div>
    </header>
<main class="content" role="main">
	<div class="container">
		<article class="post">
	<header class="post-header">
        <h3 class="p-post-title">Replacing Disqus with Github Comments</h3>
    </header>

    <section class="post-content">
        <p>I’ve been considering removing comments from this blog for a while; mainly because the site doesn’t
trigger much discussion and I didn’t like keeping the overhead of <a href="https://disqus.com/">Disqus</a>
around. After looking into Disqus load-time behaviour I was pretty shocked what I was forcing
on people loading the site (although you really should be using the likes of <a href="https://www.eff.org/privacybadger">Privacy Badger</a> and <a href="https://github.com/gorhill/uBlock">uBlock Origin</a>).</p>

<p></p>

<p>This post is Hugo-specific but the very little code required can be adapted to any website.</p>

<h5 id="what-s-wrong-with-disqus">What’s Wrong with Disqus?</h5>

<p>This is a typical request log with Disqus <em>enabled</em>:</p>

<p></p><center><a href="/img/GithubComments/DisqusLoadLogHigh.png"><img src="/img/GithubComments/DisqusLoadLog.png" alt="d"></a></center><p></p>

<p>This is the same request log with Disqus <em>disabled</em>:</p>

<p></p><center><a href="/img/GithubComments/LoadLogHigh.png"><img src="/img/GithubComments/LoadLog.png" alt="d"></a></center><p></p>

<p></p><center><strong><em>WAT!?</em></strong></center><p></p>

<p>Relevant points are:</p>

<ul>
<li>Load-time goes from 6 seconds to 2 seconds.</li>
<li>There are 105 network requests vs. 16.</li>
<li>There are a lot of non-relevant requests going through to networks that will be tracking your movements.</li>
</ul>

<p>Among the networks you can find:</p>

<ul>
<li><code>disqus.com</code> - Obviously!</li>
<li><code>google-analytics.com</code> - Multiple requests; no idea who’s capturing your movements.</li>
<li><code>connect.facebook.net</code> - If you’re logged into Facebook, they know you visit this site.</li>
<li><code>accounts.google.com</code> - Google will also map your visits to this site with any of your Google accounts.</li>
<li><code>pippio.com</code> - LiveRamp identify mapping for harvesting your details for commercial gain.</li>
<li><code>bluekai.com</code> - Identity tracking for marketing campaigns.</li>
<li><code>crwdcntrl.net</code> - Pretty suspect site listed as referenced by viruses and spyware.</li>
<li><code>exelator.com</code> - More identity and movement tracking site which even has a virus named after it!</li>
<li><code>doubleclick.net</code> - We all know this one: ad services and movement tracking, owned by Google.</li>
<li><code>tag.apxlv.net</code> - Very shady and <a href="https://better.fyi/trackers/apxlv.com/">tricky to pin-point an owner</a> as they obsfuscate their domain (I didn’t even know this was a thing!). Adds a tracking pixel to your site.</li>
<li><code>adnxs.com</code> - More tracking garbage, albeit slightly <a href="https://www.theguardian.com/technology/2012/apr/23/adnxs-tracking-trackers-cookies-web-monitoring">more prolific</a>.</li>
<li><code>adsymptotic.com</code> - Advertising and tracking that suppposedly uses machine learning.</li>
<li><code>rlcdn.com</code> - Obsfuscated advertising/tracking from Rapleaf.</li>
<li><code>adbrn.com</code> - <em>“Deliver a personalized customer journey across devices, channels and platforms with Adbrain customer ID mapping technology.”</em></li>
<li><code>nexac.com</code> - Oracle’s Datalogix, their own tracking and behavioural pattern rubbish.</li>
<li><code>tapad.com</code> - OK, I cant’t be bothered to search to look this up anymore.</li>
<li><code>liadm.com</code> - More? Oh, ok, then…</li>
<li><code>sohern.com</code> - Yup. Tracking.</li>
<li><code>demdex.net</code> - Tracking. From Adobe.</li>
<li><code>bidswitch.net</code> - I’ll give you one guess…</li>
<li><code>agkn.com</code> - …</li>
<li><code>mathtag.com</code> - Curious name, maybe it’s… no. It’s tracking you.</li>
</ul>

<p>I can’t visit many of these sites because I have them blocked in uBlock Origin so information was
gleaned from google crawl results of the webpages and 3rd parties. Needless to say, it’s a pretty disgusting insight into how certain free products turn you into
the product. What’s more worrying are the services that go to lengths to hide who they are and
what their purposes are for tracking your movements.</p>

<p>Anyway. Delete it. I’m sorry for poisoning this site with these networks. Foward to better things…</p>

<h5 id="using-github-for-comments">Using Github for Comments</h5>

<p>I was discussing removing Disqus with others and <a href="https://twitter.com/HarryDenholm">@HarryDenholm</a>
mentioned that a friend managed to get comments on their Github-hosted static blog using Pull Requests. I thought this was pretty genius and wanted to figure out if I could do something
similar for Hugo on external sites.</p>

<p>The answer came from the <a href="https://developer.github.com/v3/issues/">Github Issue API</a>. The process is quite simple and is already active on this blog post:</p>

<ol>
<li>For each blog post you make, open an Issue on some repo on Github. The Issue for this post is <a href="https://github.com/dwilliamson/donw.io/issues/1">here</a>.</li>
<li>All comments are posted directly on Github.</li>
<li>Add some Javascript to your site that reads Github’s JSON description of the Issue comments and displays them.</li>
</ol>

<p>The benefits are immediate:</p>

<ul>
<li>There is zero tracking of you as a reader to this website. Github themselves only see network read requests from unnamed IPs.</li>
<li>All comments are written in <a href="https://github.com/adam-p/markdown-here/wiki/Markdown-Cheatsheet">Markdown</a>, allowing inline code, images, lists and formatting.</li>
<li>You can use Github’s notifications to informed of responses; you don’t even have to visit this site to read and contribute to discussion.</li>
<li>Although somewhat superflous, you can also integrate <a href="https://developer.github.com/v3/reactions/">Github Reactions</a> (maybe useful on larger sites).</li>
</ul>

<p>You don’t need an API key to read Github JSON data; it’s free for all to access. The comments for
this post can be read as JSON <a href="https://api.github.com/repos/dwilliamson/donw.io/issues/1/comments">here</a>. The first comment looks like:</p>

<pre class=" language-json"><code class=" language-json">{
    "url": "https://api.github.com/repos/dwilliamson/donw.io/issues/comments/295004846",
    "html_url": "https://github.com/dwilliamson/donw.io/issues/1#issuecomment-295004846",
    "issue_url": "https://api.github.com/repos/dwilliamson/donw.io/issues/1",
    "id": 295004846,
    "user": {
        "login": "dwilliamson",
        "id": 1532903,
        "avatar_url": "https://avatars3.githubusercontent.com/u/1532903?v=3",
        "gravatar_id": "",
        "url": "https://api.github.com/users/dwilliamson",
        "html_url": "https://github.com/dwilliamson",
        "followers_url": "https://api.github.com/users/dwilliamson/followers",
        "following_url": "https://api.github.com/users/dwilliamson/following{/other_user}",
        "gists_url": "https://api.github.com/users/dwilliamson/gists{/gist_id}",
        "starred_url": "https://api.github.com/users/dwilliamson/starred{/owner}{/repo}",
        "subscriptions_url": "https://api.github.com/users/dwilliamson/subscriptions",
        "organizations_url": "https://api.github.com/users/dwilliamson/orgs",
        "repos_url": "https://api.github.com/users/dwilliamson/repos",
        "events_url": "https://api.github.com/users/dwilliamson/events{/privacy}",
        "received_events_url": "https://api.github.com/users/dwilliamson/received_events",
        "type": "User",
        "site_admin": false
    },
    "created_at": "2017-04-18T22:39:16Z",
    "updated_at": "2017-04-18T22:39:16Z",
    "body": "This is a comment"
},    
</code></pre>

<p>The first step is to add a new template to your partials directory that reads and displays Github
comments (<code>comments.html</code>). Here’s the code I use:</p>

<pre class=" language-js"><code class=" language-js"><span class="token keyword">var</span> url <span class="token operator">=</span> <span class="token string">"https://github.com/dwilliamson/donw.io/issues/"</span> <span class="token operator">+</span> <span class="token punctuation">{</span><span class="token punctuation">{</span> $<span class="token punctuation">.</span>Params<span class="token punctuation">.</span>ghcommentid <span class="token punctuation">}</span><span class="token punctuation">}</span>
<span class="token keyword">var</span> api_url <span class="token operator">=</span> <span class="token string">"https://api.github.com/repos/dwilliamson/donw.io/issues/"</span> <span class="token operator">+</span> <span class="token punctuation">{</span><span class="token punctuation">{</span> $<span class="token punctuation">.</span>Params<span class="token punctuation">.</span>ghcommentid <span class="token punctuation">}</span><span class="token punctuation">}</span> <span class="token operator">+</span> <span class="token string">"/comments"</span>

<span class="token function">$</span><span class="token punctuation">(</span>document<span class="token punctuation">)</span><span class="token punctuation">.</span><span class="token function">ready</span><span class="token punctuation">(</span><span class="token keyword">function</span> <span class="token punctuation">(</span><span class="token punctuation">)</span> <span class="token punctuation">{</span>
    $<span class="token punctuation">.</span><span class="token function">ajax</span><span class="token punctuation">(</span>api_url<span class="token punctuation">,</span> <span class="token punctuation">{</span>
        headers<span class="token punctuation">:</span> <span class="token punctuation">{</span>Accept<span class="token punctuation">:</span> <span class="token string">"application/vnd.github.v3.html+json"</span><span class="token punctuation">}</span><span class="token punctuation">,</span>
        dataType<span class="token punctuation">:</span> <span class="token string">"json"</span><span class="token punctuation">,</span>
        success<span class="token punctuation">:</span> <span class="token keyword">function</span><span class="token punctuation">(</span>comments<span class="token punctuation">)</span> <span class="token punctuation">{</span>
            <span class="token function">$</span><span class="token punctuation">(</span><span class="token string">"#gh-comments-list"</span><span class="token punctuation">)</span><span class="token punctuation">.</span><span class="token function">append</span><span class="token punctuation">(</span><span class="token string">"Visit the &lt;b&gt;&lt;a href='"</span> <span class="token operator">+</span> url <span class="token operator">+</span> <span class="token string">"'&gt;Github Issue&lt;/a&gt;&lt;/b&gt; to comment on this post"</span><span class="token punctuation">)</span><span class="token punctuation">;</span>
            $<span class="token punctuation">.</span><span class="token function">each</span><span class="token punctuation">(</span>comments<span class="token punctuation">,</span> <span class="token keyword">function</span><span class="token punctuation">(</span>i<span class="token punctuation">,</span> comment<span class="token punctuation">)</span> <span class="token punctuation">{</span>

                <span class="token keyword">var</span> date <span class="token operator">=</span> <span class="token keyword">new</span> <span class="token class-name">Date</span><span class="token punctuation">(</span>comment<span class="token punctuation">.</span>created_at<span class="token punctuation">)</span><span class="token punctuation">;</span>

                <span class="token keyword">var</span> t <span class="token operator">=</span> <span class="token string">"&lt;div id='gh-comment'&gt;"</span><span class="token punctuation">;</span>
                t <span class="token operator">+</span><span class="token operator">=</span> <span class="token string">"&lt;img src='"</span> <span class="token operator">+</span> comment<span class="token punctuation">.</span>user<span class="token punctuation">.</span>avatar_url <span class="token operator">+</span> <span class="token string">"' width='24px'&gt;"</span><span class="token punctuation">;</span>
                t <span class="token operator">+</span><span class="token operator">=</span> <span class="token string">"&lt;b&gt;&lt;a href='"</span> <span class="token operator">+</span> comment<span class="token punctuation">.</span>user<span class="token punctuation">.</span>html_url <span class="token operator">+</span> <span class="token string">"'&gt;"</span> <span class="token operator">+</span> comment<span class="token punctuation">.</span>user<span class="token punctuation">.</span>login <span class="token operator">+</span> <span class="token string">"&lt;/a&gt;&lt;/b&gt;"</span><span class="token punctuation">;</span>
                t <span class="token operator">+</span><span class="token operator">=</span> <span class="token string">" posted at "</span><span class="token punctuation">;</span>
                t <span class="token operator">+</span><span class="token operator">=</span> <span class="token string">"&lt;em&gt;"</span> <span class="token operator">+</span> date<span class="token punctuation">.</span><span class="token function">toUTCString</span><span class="token punctuation">(</span><span class="token punctuation">)</span> <span class="token operator">+</span> <span class="token string">"&lt;/em&gt;"</span><span class="token punctuation">;</span>
                t <span class="token operator">+</span><span class="token operator">=</span> <span class="token string">"&lt;div id='gh-comment-hr'&gt;&lt;/div&gt;"</span><span class="token punctuation">;</span>
                t <span class="token operator">+</span><span class="token operator">=</span> comment<span class="token punctuation">.</span>body_html<span class="token punctuation">;</span>
                t <span class="token operator">+</span><span class="token operator">=</span> <span class="token string">"&lt;/div&gt;"</span><span class="token punctuation">;</span>
                <span class="token function">$</span><span class="token punctuation">(</span><span class="token string">"#gh-comments-list"</span><span class="token punctuation">)</span><span class="token punctuation">.</span><span class="token function">append</span><span class="token punctuation">(</span>t<span class="token punctuation">)</span><span class="token punctuation">;</span>
            <span class="token punctuation">}</span><span class="token punctuation">)</span><span class="token punctuation">;</span>
        <span class="token punctuation">}</span><span class="token punctuation">,</span>
        error<span class="token punctuation">:</span> <span class="token keyword">function</span><span class="token punctuation">(</span><span class="token punctuation">)</span> <span class="token punctuation">{</span>
            <span class="token function">$</span><span class="token punctuation">(</span><span class="token string">"#gh-comments-list"</span><span class="token punctuation">)</span><span class="token punctuation">.</span><span class="token function">append</span><span class="token punctuation">(</span><span class="token string">"Comments are not open for this post yet."</span><span class="token punctuation">)</span><span class="token punctuation">;</span>
        <span class="token punctuation">}</span>
    <span class="token punctuation">}</span><span class="token punctuation">)</span><span class="token punctuation">;</span>
<span class="token punctuation">}</span><span class="token punctuation">)</span><span class="token punctuation">;</span>
</code></pre>

<p>This can then be invoked from your blog post html:</p>

<pre><code>{{ partial "comments.html" . }}
</code></pre>

<p>The variables it references need to be added to the header of your html page. In this case it’s
the single variable <code>ghcommentid</code> which is set to the number of the Issue used for comments.</p>

<h5 id="summary">Summary</h5>

<p>There’s nothing more beyond that. Technically, you could actually write your blog posts as Github
Issues themselves and forgo the need for a blog engine. But that seems too much like using a system
in ways it was never intended to be used.</p>

<p>This website is managed on Github as <a href="https://github.com/dwilliamson/donw.io">dwilliamson/donw.io</a>.
It contains the complete code for using Github for your comments.</p>
    </section>

    <hr>

    <footer class="post-footer">
        <section class="f-1">            
            
                <span class="f-post-tags"><i class="fa fa-tag"></i>
                blog
                </span>
            
            <p class="f-post-time"><time datetime="2017-04-21T20:30:00+01:00">April 21, 2017</time></p>
        </section>
                        
        <section class="f-2">
            <section class="share">
                <span>Share:
                <a class="icon-twitter" href="http://twitter.com/share?text=Replacing%20Disqus%20with%20Github%20Comments&amp;url=http%3a%2f%2fdonw.io%2fpost%2fgithub-comments%2f" onclick="window.open(this.href, 'twitter-share', 'width=550,height=235');return false;">
                    <i class="fa fa-twitter"></i>
                </a>
                <a class="icon-facebook" href="https://www.facebook.com/sharer/sharer.php?u=http%3a%2f%2fdonw.io%2fpost%2fgithub-comments%2f" onclick="window.open(this.href, 'facebook-share','width=580,height=296');return false;">
                    <i class="fa fa-facebook"></i>
                </a>
                <a class="icon-google-plus" href="https://plus.google.com/share?url=http%3a%2f%2fdonw.io%2fpost%2fgithub-comments%2f" onclick="window.open(this.href, 'google-plus-share', 'width=490,height=530');return false;">
                    <i class="fa fa-google-plus"></i>
                </a>
                </span>
            </section>
        </section>
    </footer>
    
<div id="gh-comments">
    <br><br>
    <h6>COMMENTS</h6>
    <div id="gh-comments-list"><a href="https://github.com/dwilliamson/donw.io/issues/1#new_comment_field" rel="nofollow" class="btn">Post a comment on Github</a><div id="gh-comment"><img src="https://avatars0.githubusercontent.com/u/1532903?v=4" width="24px"><b><a href="https://github.com/dwilliamson">dwilliamson</a></b> posted at <em>Tue, 18 Apr 2017 22:39:16 GMT</em><div id="gh-comment-hr"></div><p>This is a comment</p></div><div id="gh-comment"><img src="https://avatars0.githubusercontent.com/u/1532903?v=4" width="24px"><b><a href="https://github.com/dwilliamson">dwilliamson</a></b> posted at <em>Tue, 18 Apr 2017 23:01:59 GMT</em><div id="gh-comment-hr"></div><p>This is another comment</p></div><div id="gh-comment"><img src="https://avatars0.githubusercontent.com/u/1532903?v=4" width="24px"><b><a href="https://github.com/dwilliamson">dwilliamson</a></b> posted at <em>Tue, 18 Apr 2017 23:31:34 GMT</em><div id="gh-comment-hr"></div><p>This is a comment with some markdown...</p>
<ul>
<li>Listy</li>
<li>Listo</li>
</ul>
<ol>
<li>Jumble</li>
<li>Sale</li>
</ol>
<p><strong>Boldit</strong> or <em>Italicit</em> or <del>strikeit</del></p>
<h2>header</h2>
<h2>hashtag</h2>
<p>something <code>something</code></p>
<div class="highlight highlight-source-c++"><pre><span class="pl-k">int</span> x = <span class="pl-c1">3</span>;</pre></div></div><div id="gh-comment"><img src="https://avatars0.githubusercontent.com/u/1350715?v=4" width="24px"><b><a href="https://github.com/jasminpatry">jasminpatry</a></b> posted at <em>Fri, 21 Apr 2017 23:20:23 GMT</em><div id="gh-comment-hr"></div><p>This is great. Thanks for sharing this!</p></div><div id="gh-comment"><img src="https://avatars1.githubusercontent.com/u/1068290?v=4" width="24px"><b><a href="https://github.com/olafleur">olafleur</a></b> posted at <em>Sat, 22 Apr 2017 01:30:38 GMT</em><div id="gh-comment-hr"></div><p>Awesome ! I think I will try this on my website. :)</p></div><div id="gh-comment"><img src="https://avatars1.githubusercontent.com/u/35047?v=4" width="24px"><b><a href="https://github.com/vielmetti">vielmetti</a></b> posted at <em>Sat, 22 Apr 2017 01:41:20 GMT</em><div id="gh-comment-hr"></div><p>Nicely done.</p></div><div id="gh-comment"><img src="https://avatars2.githubusercontent.com/u/3889405?v=4" width="24px"><b><a href="https://github.com/vyp">vyp</a></b> posted at <em>Sat, 22 Apr 2017 01:43:24 GMT</em><div id="gh-comment-hr"></div><p>Neat!</p></div><div id="gh-comment"><img src="https://avatars3.githubusercontent.com/u/416585?v=4" width="24px"><b><a href="https://github.com/olalonde">olalonde</a></b> posted at <em>Sat, 22 Apr 2017 01:45:11 GMT</em><div id="gh-comment-hr"></div><p>You are on HN, congrats :) <a href="https://news.ycombinator.com/item?id=14170041" rel="nofollow">https://news.ycombinator.com/item?id=14170041</a></p></div><div id="gh-comment"><img src="https://avatars3.githubusercontent.com/u/7993559?v=4" width="24px"><b><a href="https://github.com/mannu1200">mannu1200</a></b> posted at <em>Sat, 22 Apr 2017 01:54:44 GMT</em><div id="gh-comment-hr"></div><p>I agree that what Disqus is doing is an overkill.<br>
However, we should not forget that nothing comes for free...<br>
Apart from this, getting redirected to GitHub for comment is very inconvenient...<br>
Update: So I am automatically subscribed to comments thread via email...<br>
Getting emails not only for replies on my comment but for every new comment.<br>
Well, this is another overkill...<br>
Don't forget User experience is more valued than user's data nowadays :D</p>
<p>Edit:<br>
<a href="https://github.com/dwilliamson" class="user-mention">@dwilliamson</a> forget about the free-ness argument, the bigger issue is of user experience, I got resubscribed to email updates after you mentioned me in one of the comment. This is a serious poor user experience.</p></div><div id="gh-comment"><img src="https://avatars1.githubusercontent.com/u/704302?v=4" width="24px"><b><a href="https://github.com/rattrayalex">rattrayalex</a></b> posted at <em>Sat, 22 Apr 2017 02:03:22 GMT</em><div id="gh-comment-hr"></div><p>Interesting, thanks for sharing!</p>
<p>You might consider changing the link to include <code>#new_comment_field</code> (<a href="https://github.com/dwilliamson/donw.io/issues/1/#new_comment_field">like so</a>) so that users do not have to scroll. I believe this also focuses the field by default, which may make intent more clear.</p>
<p>Adding a button at the bottom of the comments section that opens the link might be helpful. Using <code>about: _blank</code> could be nice too.</p>
<p>A git hook might be useful to automatically generate the issues, though I'm not sure that'd be a good idea.</p>
<p>EDIT: Looks like their api supports POSTing comments; you'd have to set up user auth of course. <a href="https://developer.github.com/v3/issues/comments/#create-a-comment">https://developer.github.com/v3/issues/comments/#create-a-comment</a></p></div><div id="gh-comment"><img src="https://avatars0.githubusercontent.com/u/951921?v=4" width="24px"><b><a href="https://github.com/ontouchstart">ontouchstart</a></b> posted at <em>Sat, 22 Apr 2017 02:10:41 GMT</em><div id="gh-comment-hr"></div><p>A test of comment with image and link.</p>
<p><a href="https://github.com/dwilliamson/donw.io/commit/d7f1f5874d704b2d68a90139e14f4a8d247e2f7b#commitcomment-21864198"><img src="https://cloud.githubusercontent.com/assets/951921/25300522/17fee0d8-26df-11e7-9588-8674b7d3b81a.PNG" alt="img_4698" style="max-width:100%;"></a></p></div><div id="gh-comment"><img src="https://avatars3.githubusercontent.com/u/505366?v=4" width="24px"><b><a href="https://github.com/globalcitizen">globalcitizen</a></b> posted at <em>Sat, 22 Apr 2017 02:20:30 GMT</em><div id="gh-comment-hr"></div><p>Good work. In summary, to improve the solution further, add a button to add a new comment. Only when that button is pushed (in order to preserve privacy, remove drive-by viewer tracking potential, and reduce unnecessary server load on Github), somehow determine whether the user is already logged in to Github.</p>
<ul>
<li>If not, then give them a chance to do so or sign up</li>
<li>If so, then display an add comment form on the page<br>
This could be a modal window, in a similar style to Stripe. If the user posts a comment successfully, the comment stream could be refreshed.</li>
</ul></div><div id="gh-comment"><img src="https://avatars3.githubusercontent.com/u/9056756?v=4" width="24px"><b><a href="https://github.com/citrusui">citrusui</a></b> posted at <em>Sat, 22 Apr 2017 02:51:30 GMT</em><div id="gh-comment-hr"></div><p>I'm impressed by this! I'm going to see what happens when you post a comment from the <a href="https://github.com/ghost" class="user-mention">@ghost</a> account (another name for a deleted account)</p></div><div id="gh-comment"><img src="https://avatars3.githubusercontent.com/u/10137?v=4" width="24px"><b><a href="https://github.com/ghost">ghost</a></b> posted at <em>Sat, 22 Apr 2017 02:53:12 GMT</em><div id="gh-comment-hr"></div><p>Hello world! This is a comment from the <a href="https://github.com/ghost" class="user-mention">@ghost</a> account.</p></div><div id="gh-comment"><img src="https://avatars1.githubusercontent.com/u/1946610?v=4" width="24px"><b><a href="https://github.com/segiddins">segiddins</a></b> posted at <em>Sat, 22 Apr 2017 02:55:22 GMT</em><div id="gh-comment-hr"></div><p>This is awesome!</p></div><div id="gh-comment"><img src="https://avatars3.githubusercontent.com/u/416585?v=4" width="24px"><b><a href="https://github.com/olalonde">olalonde</a></b> posted at <em>Sat, 22 Apr 2017 03:14:03 GMT</em><div id="gh-comment-hr"></div><p>You might want to</p>
<ol>
<li>
<p>Add <code>rel="nofollow"</code> to your anchor tags to deter spam</p>
</li>
<li>
<p>Link back to the blog post in the GH issue (you could do that in the first comment and not display it on your website). That way when I go back here due to <img class="emoji" title=":octocat:" alt=":octocat:" src="https://assets-cdn.github.com/images/icons/emoji/octocat.png" height="20" width="20" align="absmiddle"> GH notifications, I can go back to the post easily instead of wondering what this is all about <g-emoji alias="smile_cat" fallback-src="https://assets-cdn.github.com/images/icons/emoji/unicode/1f638.png" ios-version="6.0">😸</g-emoji></p>
</li>
</ol></div><div id="gh-comment"><img src="https://avatars0.githubusercontent.com/u/1504756?v=4" width="24px"><b><a href="https://github.com/ElijahLynn">ElijahLynn</a></b> posted at <em>Sat, 22 Apr 2017 03:21:03 GMT</em><div id="gh-comment-hr"></div><p>Great writeup, I was really hoping there was a way to comment without visiting Github.</p></div><div id="gh-comment"><img src="https://avatars2.githubusercontent.com/u/70218?v=4" width="24px"><b><a href="https://github.com/asciimo">asciimo</a></b> posted at <em>Sat, 22 Apr 2017 03:29:39 GMT</em><div id="gh-comment-hr"></div><p>Very slick! I came to propose something similar to <a href="https://github.com/globalcitizen" class="user-mention">@globalcitizen</a>'s proposal, but they put it more succinctly.</p></div><div id="gh-comment"><img src="https://avatars1.githubusercontent.com/u/7413661?v=4" width="24px"><b><a href="https://github.com/bexelbie">bexelbie</a></b> posted at <em>Sat, 22 Apr 2017 03:42:12 GMT</em><div id="gh-comment-hr"></div><p>Did you consider <a href="https://staticman.net">https://staticman.net</a> ? This gets you comments as PRs and comments from the page.</p></div><div id="gh-comment"><img src="https://avatars1.githubusercontent.com/u/5678977?v=4" width="24px"><b><a href="https://github.com/tbodt">tbodt</a></b> posted at <em>Sat, 22 Apr 2017 04:06:33 GMT</em><div id="gh-comment-hr"></div><p><a href="https://github.com/bexelbie" class="user-mention">@bexelbie</a> but staticman isn't as FUCKING AWESOME as this</p></div><div id="gh-comment"><img src="https://avatars3.githubusercontent.com/u/1691690?v=4" width="24px"><b><a href="https://github.com/dipakc">dipakc</a></b> posted at <em>Sat, 22 Apr 2017 04:39:19 GMT</em><div id="gh-comment-hr"></div><p>This is amazing! Thanks for sharing.</p></div><div id="gh-comment"><img src="https://avatars2.githubusercontent.com/u/4203458?v=4" width="24px"><b><a href="https://github.com/aleman">aleman</a></b> posted at <em>Sat, 22 Apr 2017 04:42:25 GMT</em><div id="gh-comment-hr"></div><p>Great!<br>
I can already think of a few use cases for this in my team</p></div><div id="gh-comment"><img src="https://avatars3.githubusercontent.com/u/1770265?v=4" width="24px"><b><a href="https://github.com/colinmegill">colinmegill</a></b> posted at <em>Sat, 22 Apr 2017 05:39:53 GMT</em><div id="gh-comment-hr"></div><p>Great idea. Thank you for this.</p></div><div id="gh-comment"><img src="https://avatars3.githubusercontent.com/u/6114?v=4" width="24px"><b><a href="https://github.com/crohr">crohr</a></b> posted at <em>Sat, 22 Apr 2017 05:51:15 GMT</em><div id="gh-comment-hr"></div><p>I agree this is a great idea. But why not write your post directly into the issue as well? You might want to checkout my post <a href="http://ebarnouflant.com/posts/4-turn-your-github-issues-into-blog-posts">Turn your GitHub issues into blog posts</a>, which implements the same idea but also let you publish right from your GitHub issue as well.</p></div><div id="gh-comment"><img src="https://avatars0.githubusercontent.com/u/814772?v=4" width="24px"><b><a href="https://github.com/bkaradzic">bkaradzic</a></b> posted at <em>Sat, 22 Apr 2017 05:57:51 GMT</em><div id="gh-comment-hr"></div><p>Love it! <g-emoji alias="+1" fallback-src="https://assets-cdn.github.com/images/icons/emoji/unicode/1f44d.png" ios-version="6.0">👍</g-emoji></p></div><div id="gh-comment"><img src="https://avatars0.githubusercontent.com/u/20718827?v=4" width="24px"><b><a href="https://github.com/lrzedzicki">lrzedzicki</a></b> posted at <em>Sat, 22 Apr 2017 06:08:30 GMT</em><div id="gh-comment-hr"></div><p>Very useful!</p></div><div id="gh-comment"><img src="https://avatars2.githubusercontent.com/u/1249760?v=4" width="24px"><b><a href="https://github.com/andreis">andreis</a></b> posted at <em>Sat, 22 Apr 2017 07:02:39 GMT</em><div id="gh-comment-hr"></div><p>Great technique that I'll definitely steal. Thanks!</p></div><div id="gh-comment"><img src="https://avatars1.githubusercontent.com/u/1449799?v=4" width="24px"><b><a href="https://github.com/mhoogkamer">mhoogkamer</a></b> posted at <em>Sat, 22 Apr 2017 07:12:48 GMT</em><div id="gh-comment-hr"></div><p>Brilliant idea!</p></div><div id="gh-comment"><img src="https://avatars1.githubusercontent.com/u/1835156?v=4" width="24px"><b><a href="https://github.com/EddieOne">EddieOne</a></b> posted at <em>Sat, 22 Apr 2017 07:19:12 GMT</em><div id="gh-comment-hr"></div><div class="highlight highlight-source-js"><pre>test <span class="pl-k">=</span> {
  something<span class="pl-k">:</span> <span class="pl-s"><span class="pl-pds">'</span>monkey<span class="pl-pds">'</span></span>,
  <span class="pl-en">function</span><span class="pl-k">:</span> <span class="pl-k">function</span>(<span class="pl-smi">arg</span>) { <span class="pl-k">return</span> arg <span class="pl-k">+</span> <span class="pl-s"><span class="pl-pds">'</span>YAY!<span class="pl-pds">'</span></span>  }
}</pre></div>
<p>Edit: We need pagination!</p></div><div id="gh-comment"><img src="https://avatars2.githubusercontent.com/u/520973?v=4" width="24px"><b><a href="https://github.com/cerisara">cerisara</a></b> posted at <em>Sat, 22 Apr 2017 07:27:49 GMT</em><div id="gh-comment-hr"></div><p>Great ! I'll try this out... is it doable with a gitlab instance ?</p></div></div>
    <a href="javascript:void(0)" id="gh-load-comments" class="btn" style="" onclick="DoGithubComments(1,2);">Load more comments</a>
</div>

<script type="text/javascript" src="http://donw.io/js/github-comments.js"></script>
<script type="text/javascript">
    DoGithubComments( 1 );
</script>

</article>
	</div>
</main>
    <footer id="site-footer">
        <div class="container">
            <a href="http://donw.io/index.xml" title="Get the RSS feed"><span class="tooltip"><i class="fa fa-rss"></i></span></a>
            <section>© <a href="http://donw.io/">Don Williamson</a> 2016 | All rights reserved</section>
        </div>
    </footer>

    <script type="text/javascript" src="http://donw.io/js/fittext.js"></script>
    <script type="text/javascript">
      $(".heading").fitText();
    </script>


    <script type="text/x-mathjax-config;executed=true">
        MathJax.Hub.Config({
        showProcessingMessages: false,
        messageStyle: "none",
          tex2jax: {
            inlineMath: [['$','$'], ['\\(','\\)']],
            displayMath: [['$$','$$'], ['\[','\]']],
            processEscapes: true,
            processEnvironments: true,
            skipTags: ['script', 'noscript', 'style', 'textarea', 'pre'],
            TeX: { equationNumbers: { autoNumber: "AMS" },
                 extensions: ["AMSmath.js", "AMSsymbols.js"] }
          }
        });
    </script>

    <script type="text/x-mathjax-config;executed=true">
      MathJax.Hub.Queue(function() {
        // Fix <code> tags after MathJax finishes running. This is a
        // hack to overcome a shortcoming of Markdown. Discussion at
        // https://github.com/mojombo/jekyll/issues/199
        var all = MathJax.Hub.getAllJax(), i;
        for(i = 0; i &lt; all.length; i += 1) {
            all[i].SourceElement().parentNode.className += ' has-jax';
        }
    });
    </script>
    <script type="text/javascript" src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.1/MathJax.js?config=TeX-AMS-MML_HTMLorMML">
    </script>




</body>
```

**Comment by deftworker** [comment]
What happens when someone decides to link sites for SEO juice off of your domain, hosting these comments?

**Comment by dwilliamson** [comment]
They already do that with my blog content: it's copied and pasted (plagiarized) many times over on horrible little websites all over the world (mostly China) with no link-back.

**Comment by darekkay** [comment]
Thanks for describing this method. I've included it in my [write-up](https://darekkay.com/blog/static-site-comments/) about different options to include comments on a static site.

**Comment by danlucraft** [comment]
Comment

**Comment by jdanyow** [comment]
Hi :wave: 
https://utteranc.es automates the blog-post to issue linking and allows users to post comments without navigating to GitHub :fire:

**Comment by shrektan** [comment]
wow, this is awesome !

**Comment by strakaps** [comment]
thanks for sharing!

**Comment by chrisgb-dev** [comment]
Fabulous idea, thanks for sharing the how too!

**Comment by zoes6** [comment]
xxxl

**Comment by rani-pinchuk** [comment]
Very nice!

**Comment by klejnov** [comment]
Work :)

**Comment by drmingdrmer** [comment]
does it work?

**Comment by peterstollenwerk** [comment]
Really?

**Comment by aldenjenkins** [comment]
testy test

**Comment by kabus202** [comment]
d

**Comment by unitedt** [comment]
Interesting idea

**Comment by LuD1161** [comment]
Great idea :+1:

**Comment by vchrombie** [comment]
Nice idea, but what if a user doesn't have a GitHub account?

**Comment by dwilliamson** [comment]
@vchrombie then they can't comment. As I mentioned, I'm not trying to build a community, only give technical people a chance to query my technical blog posts. Given the nature of my blog, there's a very high chance they have a Github account. If not and don't have the motivation to create one to ask further questions then there's no real loss for either party.

**Comment by slavafyi** [comment]
Nice!

**Comment by unklebonehead** [comment]
This is pure genius!

**Comment by wallabyway** [comment]
Very cool

**Comment by VincentTam** [comment]
@vchrombie I see you concern.  In addition, this technology only supports GitHub.

IMHO, [Staticman](https://staticman.net) gives Git(Hub/Lab) repo owners a greater degree of freedom.  Staticman's support for GitLab has been addded for three months. (c.f. mmistakes/minimal-mistakes#1956)

**Comment by dewomser** [comment]
Test this and Staticman and others. **bold things**
1.you
2.never
3.know …
@dewomser maintains [wolust.de](https://wolust.del)

**Comment by clarkttfu** [comment]
Awesome !!

**Comment by iamskok** [comment]
@bdorlov Hey, there! Just testing out github comments!

**Comment by HyeonGuJ** [comment]
wow!

**Comment by dvijayak** [comment]
I think this is super cool. I plan on implementing something like this on [my own site](https://dvijayak.net) in due time.

**Comment by ingochris** [comment]
`Hello world! 
`
- Yet another test comment. :)

1. Do images display? 
![snoo](https://raw.githubusercontent.com/redditbwf/redditbwf.github.io/master/assets/images/bwfsnoo2.png)

Cheers!
Chris

**Comment by iamskok** [comment]
XSS test. Open console.

<img src="https://avatars.mds.yandex.net/get-pdb/28866/4f80988c-b9d8-4a18-9e93-4ad8698af842/s1200"  onload="console.log('XSS from github comments')"/>

**Comment by yudiqing** [comment]
This is a terrific post. Thanks.

**Comment by pahosler** [comment]
Flutter Slide Puzzle Demo [Repo](https://github.com/pahosler/Flutter-Slide-Puzzle)
<img src="https://github.com/pahosler/Flutter-Slide-Puzzle/raw/master/images/demo.gif" width="250">

**Comment by ghost** [comment]
Thanks for sharing!

**Comment by anonizmous** [comment]
OhOhOh....!!!

**Comment by JadedBlueEyes** [comment]
BTW guys, you may like http://utteranc.es that does the same thing, but a bit less hackery.

**Comment by SergiiVdovareize** [comment]
и как оно?

**Comment by begoon** [comment]
вроде работает

**Comment by drom** [comment]
да, хорошая идея

**Comment by isovector** [comment]
This is an amazing idea!

**Comment by craftsmandigital** [comment]
Hello folks

**Comment by Noobchenko** [comment]
Проверка как работает

**Comment by md2perpe** [comment]
[Code](https://help.github.com/en/articles/creating-and-highlighting-code-blocks) test:
```javascript
let test = "Just a string";
another(line);
yet.another(line);
```

**Edit**: It seems like some extra CSS is needed for the code colorization.

**Comment by magicxavi** [comment]
haha

**Comment by adriansinger87** [comment]
This is a cool way to use GH for both, as a host for static jekyll-blogs and for its comments. I will try this out!

**So thank you for your work**, here is my implementation: https://www.adrian-singer.de

**Comment by cphthomas** [comment]
Very nice post :O)

**Comment by ajbozdar** [comment]
It's interesting. Cheers!

**Comment by gavinft** [comment]
This is very cool

**Comment by ldoublewood** [comment]
this is my comment

**Comment by p4ttch** [comment]
this is pretty neet!

**Comment by resthyphen** [comment]
> this is pretty neet!

agree

**Comment by benstaf** [comment]
i am trying this github comment feature

**Comment by dollardhingra** [comment]
amazing man

**Comment by pedroth** [comment]
Great idea maybe will do the same

**Comment by Feelcame** [comment]
fdfdfdfd

**Comment by hadooper01** [comment]
chcking if live comment

**Comment by hadooper01** [comment]
will try to add this on my tech blog [https://geekysoul.com/](Geekysoul) . Thanks a ton!

**Comment by rioastamal** [comment]
# test markdown for h1

**Comment by liuteriadautore** [comment]
Well, so you mean that I can post a "comment like that" right here?

**Comment by vuhanova** [comment]
comment

**Comment by BK1031** [comment]
cool comment

**Comment by rezazsouilah** [comment]
good!

**Comment by x2v0** [comment]
**Test**
#
# eee

**Comment by jonathanvanderhout** [comment]
Test. 

That's rad.

**Comment by sbrooker3** [comment]
this is another comment

**Comment by latenightly** [comment]
what a funny idea, i just wanted to try it out. I might use it myself. Thank you for making me randomly find you somewhere in the internet. that's why i love it.

**Comment by latenightly** [comment]
> what a funny idea, i just wanted to try it out. I might use it myself. Thank you for making me randomly find you somewhere in the internet. that's why i love it.

@latenightly well i am self-referencing, but just trying out. Maybe i should tag someone else like @sbrooker3 just for this test. Lol?

**Comment by melbis** [comment]
testing

**Comment by bloggrammer** [comment]
This is great to know. How about nontechie without a GitHub account?

**Comment by dwilliamson** [comment]
@Blogrammer I answered that one here https://github.com/dwilliamson/donw.io/issues/1#issuecomment-423857747

**Comment by neshant** [comment]
good stuff! thanks for a detailed blog post

**Comment by jscreatordev** [comment]
comment

**Comment by ond1** [comment]
test

**Comment by voightachrome** [comment]
thanks alot man

**Comment by voightachrome** [comment]
test test

**Comment by Tom-Gooding** [comment]
Awesome, thanks for sharing this!

**Comment by ozh** [comment]
Thanks !
